### PR TITLE
task(settings): Finish model / validation pattern

### PR DESCRIPTION
**Pull request by dschom**
## Because

- We want to use class-validator for more standardized validation support
- We want to the link validator to leverage model validation

## This pull request

- Applies class-validator to all models
- Makes minor tweaks to the bind-decorator to support class-validator
- Makes minor tweaks to model-data-store to support class-validator
- Adds DIP to
- Ensures undefined is returned for unspecified values
- Cleans up link validator to use model.validate().
- Removes direct access to underlying model data used by ModelDataProvider
- Adds dirty flag to ModelDataProvider to reduce unnecessary validations
- Revisits all validators and attempts to reach parity with content-server

## Issue that this pull request solves

Closes: FXA-8099

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Other information (Optional)

I did my best to reach parity with model validation in content server. That being said there are a couple `// TODO - Validation -` comments in the code that are probably worth revisiting. Perhaps a follow up is in order. Further more, we don't have our validation in settings under test, so this is worth filing a follow up for as well.

### CODE REVIEW

File: packages/fxa-settings/package.json
Comment (LZoog) [inline packages/fxa-settings/package.json:81]: Following up on [this comment](https://github.com/mozilla/fxa/pull/15493#issuecomment-1631674385), any other pros/cons worth mentioning here? Anything you gleaned while working on this can help inform FXA-8100.

File: packages/fxa-settings/src/lib/integrations/integration-factory.ts
Comment (LZoog) [inline packages/fxa-settings/src/lib/integrations/integration-factory.ts:187]: Just so I'm clear, when should we use `getModelData` vs referring to `integration.data` directly?

File: packages/fxa-settings/src/models/integrations/oauth-integration.ts
Comment (LZoog) [inline packages/fxa-settings/src/models/integrations/oauth-integration.ts:63]: Something I noticed recently is in the backbone version, if client_id is present, I think we always try to fetch/validate it immediately. If you go to an RP in staging and change `client_id`, it immediately shows an "invalid param client_id" error, and in the React version we don't immediately error and probably should. Not sure if we can read from the config here to check against valid client_ids but feels like that's another task, maybe FXA-8106 or FXA-6836, but if you're going to file a follow up for the other "TODO - Validation" comments I could see us looking at it then.

**Review by LZoog — APPROVED** [review]
LGTM! Had some follow up convos in Slack about a couple things, LMK if you'd like me to look again before merge.

File: packages/fxa-settings/package.json
Comment (dschom) [reply packages/fxa-settings/package.json:81]: I think there's a few points:

- The class validator gives us a lot of bang for the buck and paired well the the `@bind` decorator
- We can still validate on read/write access via some trickery in the model provider
- We can write custom validations if needed to reduce repetition. Personally I like the obvious nature of just using the standard validation set they provided.
- The error messaging and context for the validator is pretty nice

One downside:
- The entire model has to be validated. You can't target a specific field. There are 'validation groups' to validate a subset of a class's fields, but this is pretty noisy to apply at a granular level or a field by field basis.

File: packages/fxa-settings/src/lib/integrations/integration-factory.ts
Comment (dschom) [reply packages/fxa-settings/src/lib/integrations/integration-factory.ts:187]: We can still use `intergration.data` directly; however, we can't get direct access to its data store anymore.

The reason behind this is that we now track if there are changes to the model's data source via an `isDirty` flag. The `isDirty` flag let's us avoid unnecessary model validations. The only way to manage this flag is to encapsulate the object that stores the data so we know when there are changes to it. I'll leave a call out on this change so that this comment makes a bit more sense.

Essentially `integration.data` has two new methods. `getModelData(key)` and `setModelData(key,val)`, and `getModelData()` which provided direct access to the model's data store was removed.

File: packages/fxa-settings/src/models/integrations/oauth-integration.ts
Comment (dschom) [reply packages/fxa-settings/src/models/integrations/oauth-integration.ts:63]: That is interesting case. Thanks for call it out. I think we probably could do this dynamically. All though so far all the validations are statically defined. I've added the `TODO - Validation` comment here.

**Pull request merged**
